platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -51,12 +54,20 @@ public abstract class AbstractCollectionProperty<T, C extends Collection<T>> ext
         init(defaultValue, noValueSupplier());
     }
 
-    private CollectionSupplier<T, C> emptySupplier() {
-        return new EmptySupplier();
+    private CollectionSupplierGuard<T, C> emptySupplier() {
+        return guardSupplier(new EmptySupplier());
+    }
+
+    private CollectionSupplierGuard<T, C> noValueSupplier() {
+        return guardSupplier(Cast.uncheckedCast(NO_VALUE));
+    }
+
+    private void setSupplier(CollectionSupplier<T, C> unguardedSupplier) {
```

Comment:
It seems this produces two levels of guards, for instance (or maybe specifically), when invoked from `addCollector()`. 

Should a helper method that only wraps a collection supplier into a `CollectionSupplierGuard` if the given collection supplier is not a `CollectionSupplierGuard` itself add any noticeable overhead?

Reply:
I take it back, please ignore me.